morl_baselines/single_policy/ser/mo_q_learning.py

```diff
@@ -138,22 +159,43 @@ def update(self):
         if next_obs not in self.q_table:
             self.q_table[next_obs] = np.zeros((self.action_dim, self.reward_dim))
 
-        max_q = self.q_table[next_obs][self.eval(self.next_obs)]
+        max_q = self.q_table[next_obs][self.eval(self.next_obs, self.weights)]
         td_error = self.reward + (1 - self.terminated) * self.gamma * max_q - self.q_table[obs][self.action]
         self.q_table[obs][self.action] += self.learning_rate * td_error
 
         # Dyna updates
         if self.dyna:
-            self.model.update(obs, self.action, self.reward, next_obs, self.terminated)
+            if self.gpi_pd:
+                priority = (
+                    np.dot(self.reward, self.weights)
+                    + (1 - self.terminated) * self.gamma * self.parent.max_scalar_q_value(self.next_obs, self.weights)
+                    - np.dot(self.q_table[obs][self.action], self.weights)
+                )
+                priority = max(np.abs(priority), self.min_priority) ** self.alpha
+            else:
+                priority = None
+
+            self.model.update(obs, self.action, self.reward, next_obs, self.terminated, priority)
             for _ in range(self.dyna_updates):
-                s, a, r, next_s, terminal = self.model.random_transition()
+                if self.gpi_pd:
+                    s, a, r, next_s, terminal, ind = self.model.random_transition()
+                else:
+                    s, a, r, next_s, terminal = self.model.random_transition()
                 if s not in self.q_table:
                     self.q_table[s] = np.zeros((self.action_dim, self.reward_dim))
                 if next_s not in self.q_table:
                     self.q_table[next_s] = np.zeros((self.action_dim, self.reward_dim))
-                max_q = self.q_table[next_s][self.eval(next_s)]
+                max_q = self.q_table[next_s][self.eval(next_s, self.weights)]
                 model_td = r + (1 - terminal) * self.gamma * max_q - self.q_table[s][a]
                 self.q_table[s][a] += self.learning_rate * model_td
+                if self.gpi_pd:
+                    priority = (
```

Comment:
I think we can make a function to compute the priority to avoid code duplication with line 170 here.